### PR TITLE
[BUG] Introduce parser supplier support in FileSystemDocumentLoader

### PR DESCRIPTION
**Pull request by KaisNeffati**
## Issue

https://github.com/langchain4j/langchain4j/issues/1026

## General checklist

- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [X] I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green

- [X] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)

## Checklist for adding new model integration

- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for adding new embedding store integration

- [ ] I have added a `{NameOfIntegration}EmbeddingStoreIT` that extends from either `EmbeddingStoreIT` or `EmbeddingStoreWithFilteringIT`
- [ ] I have added my new module in the [BOM](https://github.com/langchain4j/langchain4j/blob/main/langchain4j-bom/pom.xml)

## Checklist for changing existing embedding store integration

- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

### CODE REVIEW

**Comment by langchain4j** [comment]
@KaisNeffati thanks a lot!

I am not sure the proposed solution will solve the issue, as:
- Current users already use existing methods of `FileSystemDocumentLoader` and they will have to change their code to use new methods with suppliers in order to fix the issue
- New users will not necessarily find and use the "correct" new methods with suppliers, so the wrong behavior might still persist for them and be unobvious
- All other parsers are stateless, so having `ApacheTikaDocumentParser` statefull is counter-intuitive

Instead, I would propose to make `ApacheTikaDocumentParser` stateless:
- the creation of `Parser`, `ContentHandler` and other components should be done in `parse` method instead of the constructor
- a new constructor should be added. this ctor should accept suppliers of `Parser`, `ContentHandler` and other tika components
- existing non-default constructor should be marked as `@Deprecated` with a comment to use ctor with suppliers instead if user wants to use this parser for multiple files

WDYT? Thanks!

**Comment by KaisNeffati** [comment]
Yeah ! that's a better approach , I'll update the PR accordingly !

Well seen !

File: document-parsers/langchain4j-document-parser-apache-tika/src/main/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParser.java [hunk 1]
```diff
@@ -48,22 +53,45 @@ public ApacheTikaDocumentParser() {
...
+        this.metadataSupplier = metadataSupplier;
+        this.parseContextSupplier = parseContextSupplier;
     }
 
     // TODO allow automatically extract metadata (e.g. creator, last-author, created/modified timestamp, etc)
 
     @Override
     public Document parse(InputStream inputStream) {
         try {
+            Parser parser = getOrDefault(parserSupplier, () -> DEFAULT_PARSER_SUPPLIER).get();
```
Comment (langchain4j) [inline document-parsers/langchain4j-document-parser-apache-tika/src/main/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParser.java]: minor: I would call `getOrDefault` in the constructor instead and do `parserSupplier.get()` here (same for the rest of fields)

File: document-parsers/langchain4j-document-parser-apache-tika/src/test/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParserTest.java [hunk 2]
```diff
@@ -40,7 +41,7 @@ void should_parse_doc_ppt_and_pdf_files(String fileName) {
     })
     void should_parse_xls_files(String fileName) {
 
-        DocumentParser parser = new ApacheTikaDocumentParser(new AutoDetectParser(), null, null, null);
+        DocumentParser parser = new ApacheTikaDocumentParser(AutoDetectParser::new, null, null, null);
```
Comment (langchain4j) [inline document-parsers/langchain4j-document-parser-apache-tika/src/test/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParserTest.java:44]: why is this change required?

**Review by ghost — COMMENTED** [review]
@KaisNeffati thank you, good job! Just a few minor comments

File: document-parsers/langchain4j-document-parser-apache-tika/src/main/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParser.java (hunk 1 above)
Comment (KaisNeffati) [reply document-parsers/langchain4j-document-parser-apache-tika/src/main/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParser.java]: Done !

File: document-parsers/langchain4j-document-parser-apache-tika/src/test/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParserTest.java (hunk 2 above)
Comment (KaisNeffati) [reply document-parsers/langchain4j-document-parser-apache-tika/src/test/java/dev/langchain4j/data/document/parser/apache/tika/ApacheTikaDocumentParserTest.java:44]: Well spotted, not required ! 👍

**Review by ghost — APPROVED** [review]
@KaisNeffati thank you! Good job!
